### ISSUE TITLE
Pointing Debug Build Packages to Sources

### ISSUE THREAD

**Issue by jandrewmc** [conan-io/conan#4736]
May be related to: Command that fetches sources for debugging #4289

To help us debug your issue please explain:

- [X] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

Conan v1.13, Centos 7, Qt and qmake, GDB

- [X] I've specified the Conan version, operating system version and any tool that can be relevant.

Our workflow consists of pulling down our source from GitHub, building our libraries, and exporting the headers, sources, and binaries as a Conan package.  We package the binaries to reduce our teams build times (some libraries take over 30 minutes to build) then we share them with the rest of the team.  The problem we are running into is related to packaging debug builds and how GDB ties out source paths.  See: [Source Path](http://sourceware.org/gdb/onlinedocs/gdb/Source-Path.html).

When we create and upload our package, GDB doesn't know how to tie out our source files meaning we cannot use GDB to step through.  Even if we used Conan to build our library once and then uploaded the package with the binaries, the source file probably will not tie out (due to different users as part of the /home/user/.conan/... path or a custom Conan user directory.).  From my research, rpm uses a tool called debugedit to modify the debug info to point the binary files to the correct sources.  

What would be really nice is that when a Conan debug package is pulled from the remote repository, somehow, someway we can tie out the binary file debug info to the sources in the package, wherever the package may reside.  

My current thought is using debugedit to change the DW_AT compilation directory on package export and either preserving the previous package location in cache and making that part of the package or using a placeholder upon upload and replacing it on install.

Further reading: [How to Point GDB to your Sources](https://alex.dzyoba.com/blog/gdb-source-path/)

- [X] I've explained the steps to reproduce the error or the motivation/use case of the question/suggestion.

**Comment by lasote** [comment]
Hi! thanks for the info. 
This is an open issue but it is in our roadmap, but we haven't managed yet to provide a solution.
I tried some months ago to use the [gdbinit](http://man7.org/linux/man-pages/man5/gdbinit.5.html) file to map to the sources of the packages. For that, I created a generator "gdb" that generates a gdbinit file for me. I can't recall very well, I didn't finish it because I found some issues but I think it is doable.

**Comment by portusato** [comment]
To be clear: there's currently no simple way to debug a package installed with conan? (of course I'm expecting the debugger to find the source files) At my company we're getting started with conan and we totally love it, but this is a feature that we need. There's no way our developer teams can use conan for distributing packages internally if these packages can't be debugged.

**Comment by lasote** [comment]
I think in Windows if you build the package you want to debug from sources, the build folder in the conan cache will contain the pdbs and the debugger will be able to locate everything. 
I haven't tried again in Linux since my attempts with the gdb. Which operating system are you using? Are you using an IDE? Which Build System? We can give it a try again.

**Comment by sourcedelica** [comment]
Maven supports this really well.  You opt-in by adding a `maven-source-plugin` to your pom and then it automatically creates a `xxx-sources.jar` when the project is built and adds jar to the artifacts for that project.

Ideally there would be a similar opt-in mechanism for Conan whereby the `sources` directory would get zipped up and added to the package on the creation side.  On the consumption side there would need to be a new generator or an addition to an existing generator to make the sources available.  The CLion Conan plugin should also be updated to support it so you could automatically jump to sources in the IDE.

This is one of the features that's holding us back from using Conan for internal packages.   We're using Conan for third-party packages and I've had to teach developers how to find the sources on Github, etc. and manually attach them to the debugger.

**Comment by noverby** [comment]
@sourcedelica @lasote 
I created the following solution for my company. 
They could be added as Conan hooks.
### Setup recipe source mapping:
```python
def package(self):
    for ext in ("c", "cpp", "cpp", "h", "hpp", "hxx"):
        conanfile.copy("*." + ext, "src")

def package_info(self):
    self.env_info.CFLAGS.append("-fdebug-prefix-map=%s=%s" % (self.source_folder, self.name))
    self.env_info.CXXFLAGS.append("-fdebug-prefix-map=%s=%s" % (self.source_folder, self.name))
    self.env_info.SOURCE_MAP.append("%s|%s" % (self.name, os.path.join(self.package_folder, "src")))
```
### Setup gdb with `.gdbinit` to use the source mapping:
```python
python
import os
if "SOURCE_MAP" in os.environ:
    for map in os.environ["SOURCE_MAP"].split(":"):
        if not "|" in map:
            continue
        gdb.execute("set substitute-path %s %s" % tuple(map.split("|")))
end
```

**Comment by lasote** [comment]
@noverby That is very interesting. I think Conan could automate/help with it, we will study it. THANKS!

**Comment by lasote** [comment]
@noverby I have a question. The `-fdebug-prefix-map` shouldn't be adjusted while compiling the sources of the package? Declaring it in the CFLAGS at the package_info will only affect to the depending packages.

**Comment by noverby** [comment]
@lasote you are right. CFLAG and CXXFLAG should only be set when compiling and not be exposed in `package_info`. We are actually not using the code I represent in this thread. The presented code was made to demonstrate how it works conceptually. Instead, we set the environment with os.environ.

**Comment by noverby** [comment]
@lasote only the SOURCE_MAP environment variable needs to be exposed and it should only happen for packages with debug information. E.g debug build or release debug build.

**Comment by noverby** [comment]
@lasote I did a fast rewrite to a Conan hook: (Not tested)
```
import os


def env_prepend(var, val, sep=os.pathsep):
    os.environ[var] = val + (sep + os.environ[var] if var in os.environ else "")

def pre_build(output, conanfile, **kwargs):
    assert conanfile
    # Set debug prefix flags
    if not (
        hasattr(conanfile, "source_folder")            # Needs source directory
        and "build_type" in conanfile.settings.fields  # Needs build_type setting
        and conanfile.settings.build_type == "Debug"   # Only for Debug build_type
    ):
        return
    env_prepend(
        "CFLAGS",
        "-fdebug-prefix-map=%s=%s" % (conanfile.source_folder, conanfile.name),
        " ",
    )
    env_prepend(
        "CXXFLAGS",
        "-fdebug-prefix-map=%s=%s" % (conanfile.source_folder, conanfile.name),
        " ",
    )

def post_package(output, conanfile, conanfile_path, **kwargs):
    assert conanfile
    # Copy sources to package
    if not (
        hasattr(conanfile, "source_folder")            # Needs source directory
        and "build_type" in conanfile.settings.fields  # Needs build_type setting
        and conanfile.settings.build_type == "Debug"   # Only for Debug build_type
    ):
        return
    for ext in ("c", "cpp", "cpp", "h", "hpp", "hxx"):
        conanfile.copy("*." + ext, "src")

def pre_package_info(output, conanfile, reference, **kwargs):
    assert conanfile
    # Set source mapping env var
    if not (
        hasattr(conanfile, "source_folder")            # Needs source directory
        and "build_type" in conanfile.settings.fields  # Needs build_type setting
        and conanfile.settings.build_type == "Debug"   # Only for Debug build_type
    ):
        return
    conanfile.env_info.SOURCE_MAP.append("%s|%s" % (conanfile.name, os.path.join(conanfile.package_folder, "src")))
```

**Comment by lasote** [comment]
A quick clarification/update @portusato @sourcedelica , for a simple solution, following the same approach as Visual Studio, if you build from sources a package you want to debug, the debugger will be able to locate the sources without any issue and without requiring further actions. I tried with gdb and with the `clion` IDE.  That works because:

1. The sources are there, in the Conan cache.
2. The hard coded paths in the libraries of the conan package point to your local cache sources.

It doesn't work without building locally the package because the hardcoded paths from the building machine (ci or any other) won't necessarily match with your local file system structure. That's where solutions like the proposed by @noverby can help.

I think we should document these issues/solutions properly in the conan docs.

**Comment by sourcedelica** [comment]
Thanks @noverby and @lasote!  This sounds promising.

**Comment by memsharded** [comment]
Closing this ticket.

The current documented simple approach is the one in https://docs.conan.io/2/examples/dev_flow/debug/step_into_dependencies.html, basically building from sources in the cache in Debug mode works easily.

There is also documentation specific for VS and PDBs here: https://docs.conan.io/2/examples/dev_flow/debug/debugging_visual.html

For most advanced uses cases, users can customize their hooks to handle differently their sources, their PDBs, etc. For very sporadic usage of PDBs, they could be added to package metadata files, and only retrieved explicitly when desired, to not have to pay the cost of always downloading them.

Please open new tickets taking those current docs in mind if necessary, thanks for the feedback.
